### ISSUE TITLE
Architecture for Piece classes

### ISSUE THREAD

**Issue by kalekundert** [kxgames/cherts#1]
I had the thought that it might be nice to share moves/attacks between pieces.  This leads to an architecture like the following, where pieces aren't subclassed, but are instead instantiated with different combinations of `Move`, `Attack`, etc. objects:

```py
class Piece:

    def __init__(self):
        self.pos = Vector()
        self.moves = []
        self.attacks = []
    ...

class Move:

    def get_possible_jumps(self, piece):
        """
        Return a list of coordinates that the piece could jump to.  The 
        player would pick which move to actually make.
        """
        return []

    def get_possible_paths(self, piece):
        """
        Return a list of paths that the piece could smoothly move along.  Each 
        path would specify a series of waypoints and a speed at which to move 
        between those waypoints.  The player would pick which path to actually 
        use.
        """
        return []

class Path:

    def __init__(self):
        self.waypoints = []
        self.speed = 0
        self.loop = False

    def get_coord(self, dt):
        """
        Return the coordinate for a piece moving along this path after the 
        given amount of time.
        """
        ...

class KnightMove(Move):
    """
    Make L-shaped jumps like a knight.
    """

    def get_possible_jumps(self, piece):
        return [
                piece.pos + (2, 1),
                piece.pos + (1, 2),
                ...
        ]

class LinearMove(Move):
    """
    Move smoothly in either the horizontal or vertical direction.
    """

    def get_possible_paths(self, piece):
        return [
                Path(...),
                Path(...),
        ]
```

This would make it easy to define pieces from a config file:

```py
def load_pieces(config_path):
    pieces = {}
    config = toml.load(config_path)

    for name in config:
        pieces[name] = p = Piece()
        p.moves = [load_move(x) for x in config[name]['moves']]

    return pieces

def load_move(config):
    if config['type'] == 'knight':
        return KnightMove(config['step_size'], ...)
    if config['type'] == 'linear':
        return LinearMove(config['speed'], config['loop'], ...)
    ...
```

**Comment by alexmitchell** [comment]
I like it.

I know this is just pseudocode, but for clarification, get_possible_jumps or get_possible_paths would be called by something external to the Piece class? Such as by the gui or ai. The Piece class is more of a passive container for piece information than an active agent (e.g. makes changes to the World). I'm asking because I'm wondering if the Move class and child classes should return relative motions which the piece to use to generate absolute motions based on current position (and facing direction?). Then the Move classes would not need to know anything about the piece. Honestly, I don't know if it will make a difference either way. ( selected_piece.get_possible_jumps() vs get_possible_jumps(selected_piece) )

I'm not familiar with toml, but it sounds good for the config file

**Comment by kalekundert** [comment]
Right, I imagine that `Piece` would be more of a passive container.  

That's a good thought about absolute vs. relative coordinates.  I do think it makes sense for `get_possible_paths()` to *return* absolute coordinates, because relative coordinates could get confusing as the piece moves along the path.  But we think of moves in relative terms, so it makes sense to organize the code that way, too.  One potenial drawback is that it might be annoying to have a bunch of  `AbsolutePath` and `RelativePath` classes with similar code, so we'll have to find a way to avoid that.

Come to think about it, we should probably load the moves from a config file too.  The way I'm thinking about it, there are basically two kinds of moves: jumps and paths.  Jumps are just a list of possible landing points, and paths are just a list of waypoints with a speed.  Both could be specified in a config file pretty easily, and that would make it even easier to experiment with different piece types.

Loading from a config file kinda solves the absolute/relative problem too, because the moves would be specified in the config file as relative paths, and then be converted to absolute paths by the piece upon request.

I like `toml`, it's my go-to config file format.
